Updating Route filter lines to take IpWildcards and other fixes

haverma opened the pull request:
- update route filter lines to take `IpWildcard`
- fix `Route4FilterLineThrough` to use the `throughPrefix` which it was ignoring earlier

anothermattbrown commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java, line 851 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpkIIT21OZzX-mE2ba:-LCpuBf58SMf-Imv3s8A:b-hcuw8b) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java#L851)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

This will result in a crash when the wildcard is not equal to a prefix. Instead, the code that uses this needs to handle IpWildcard instead of Prefix. @anothermattbrown  may be able to help here.
This applies multiple places in the `symbolic` package.
</blockquote></details>

We discussed this. Wildcards are one of many things `symbolic` doesn't currently handle. Adding support for them is out of scope for this PR. All I ask for now is that it fail loudly (throw an exception) on encountering an unsupported object, which `toPrefix` does. So this seems reasonable to me.

We might want to draw more attention to the fact that something is not handled here by doing an explicit check and throw (like we do elsewhere in `symbolic`) rather than letting `toPrefix` throw.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java, line 851 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpkIIT21OZzX-mE2ba:-LCq0xqS1betggaBHUZv:bpppci6) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java#L851)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

We discussed this. Wildcards are one of many things `symbolic` doesn't currently handle. Adding support for them is out of scope for this PR. All I ask for now is that it fail loudly (throw an exception) on encountering an unsupported object, which `toPrefix` does. So this seems reasonable to me.

We might want to draw more attention to the fact that something is not handled here by doing an explicit check and throw (like we do elsewhere in `symbolic`) rather than letting `toPrefix` throw.
</blockquote></details>

OK. I'm fine with no change, or louder exception.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 13 of 34 files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/batfish/batfish/1426#-LCpekBBE4vgHFA4pLV1:-LD8HXEuA3VKEsr4_Hn7:bd32bgr) (no related file):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Placeholder for cache fix we discussed offline
</blockquote></details>

Included the change for using `Supplier` for permitted and denied `cache` in `Route(6)FilterList`

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java, line 16 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpGHmi9rW-sk0hRkI-:-LD7zrI2CHYf8BRMWxBY:bc1uwqb) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java#L16)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I'm not sure that silently ignoring things that are not prefixes is the way to go here. What are we even using this for?
</blockquote></details>

It is unused right now,  throwing exception when `IPWildcard` cannot be converted to `Prefix`

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Route6FilterLine.java, line 91 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpGqJWBcEqc28ZkGYu:-LD4WmM4Cp5g4TNRJmkK:b-hyhc4h) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Route6FilterLine.java#L91)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Who uses this? Can we remove it?
Same applies to IPV4 version
</blockquote></details>

`RouteFilterLists.toCompactString` is being used by `RouteFilterListsDiff.genDiffInfo()` https://github.com/batfish/batfish/blob/954d6bb94365b4383568b7c1ac0d2fcf54580ada/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RouteFilterListsDiff.java#L37, which is subsequently used by the NodesQuestionPlugin. `Route6FilterLists.toCompactString` is not being used right now but I think it may get used once we start understanding IPv6

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Route6FilterLine.java, line 100 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpGrWXB9McoIQK3_BL:-LD4fGCOAQh3a3A6KBRY:bh0ag6) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Route6FilterLine.java#L100)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Seems like a fine candidate for `MoreObjects.toStringHelper`
Same applies to IPV4 version
</blockquote></details>

replaced with `MoreObjects.toStringHelper`

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/Route6FilterListTest.java, line 57 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpesK_0ofxb31EnxC9:-LD8-K5g4G-gJR4EmuaW:b-zclr5q) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/Route6FilterListTest.java#L57)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Can you use a matcher for this similar to what we have for `IpAccessListMatchers.accepts` / `IpAccessListMatchers.rejects`?
Applies a bunch of places.
</blockquote></details>

using matchers in `RouteFilterListTest` and `Route6FilterListTest`

---

*[projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java, line 2057 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpira8AIrvJsBnTpa_:-LD84xnXCCZSTtrOlQHU:b-4c2edz) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java#L2057)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

This will cause NPE if `ctx.IP_ADDRESS()` is `null`. Same below.
</blockquote></details>

checking the correct variable for not null

---

*[projects/batfish/src/main/java/org/batfish/representation/juniper/Route4FilterLineAddressMask.java, line 11 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpign044HXSRFfMCoM:-LD863rt-4xKExtA0Lgf:b-7tfmjp) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/representation/juniper/Route4FilterLineAddressMask.java#L11)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Javadoc please
</blockquote></details>

added

---

*[projects/batfish/src/main/java/org/batfish/representation/juniper/Route4FilterLineAddressMask.java, line 31 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpjkRLBn9g2FMpD5A2:-LD864lxCDyPYvlXTl-E:b-ijq3al) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/representation/juniper/Route4FilterLineAddressMask.java#L31)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Can you not combine a length constraint with an address-mask? If so, perhaps we need to restructure all of our `Route4FilterLine` subclasses. That is, a line would most generally have an address constraint and length constraint, similar to the data model version.
</blockquote></details>

no, it address mask cannot be combined with a length-constraint. Verified on GNS3
![Screen Shot 2018-05-21 at 5.54.01 PM.png](https://files.reviewableusercontent.io/images/reviewable/15151202/ojL8EhODPIJVgxldQgOpOUSH/Screen%20Shot%202018-05-21%20at%205.54.01%20PM.png)

---

*[projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java, line 851 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpkIIT21OZzX-mE2ba:-LD8BTtT3Q2SPf7OZl08:b-7d96n3) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/main/java/org/batfish/symbolic/bdd/TransferBDD.java#L851)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

OK. I'm fine with no change, or louder exception.
</blockquote></details>

Changed to a louder exception

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 217 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCpki2x8C7pX4TCW0Sp:-LD8CNBz2nr7u5LZBth3:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L217)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

If there is a name clash preventing a static import of `hasLines`, import IpAccessListMatchers so you don't need the leading `org.batfish.datamodel.matchers`. Applies in multiple places.
</blockquote></details>

done

---

*[projects/batfish/src/test/resources/org/batfish/grammar/juniper/testconfigs/route-filter, line 7 at r1](https://reviewable.io/reviews/batfish/batfish/1426#-LCplDRn8y0crQfyvQUw:-LD88UtoClzCMbvoCLvk:bw0essp) ([raw file](https://github.com/batfish/batfish/blob/eb72013c7106aee83a65c4023b6753b1605359d6/projects/batfish/src/test/resources/org/batfish/grammar/juniper/testconfigs/route-filter#L7)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Can this last line also have a length restriction like the above lines?
Same question for your `routing-policy` test
</blockquote></details>

No, addressed in a previous comment

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 21 of 22 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java, line 20 at r2](https://reviewable.io/reviews/batfish/batfish/1426#-LD8b6az8EZvT7g7RLWY:-LD8b6b-3pWsmB-z85wn:b-b30reu) ([raw file](https://github.com/batfish/batfish/blob/90ab55ff451f27f41970ff0eab85237d034c51ea/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java#L20)):*
> ```Java
>             String.format(
>                 "Cannot convert RouteFilterLine with IP Wildcard %s to PrefixSpaceLine",
>                 rfLine.getIpWildcard().isPrefix()));
> ```

Your format string will provide a substitution of `true` or `false`. Was that your intention?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/Route6FilterListMatchersImpl.java, line 49 at r2](https://reviewable.io/reviews/batfish/batfish/1426#-LD8c0BQDx15ssStxMUw:-LD8c0BRExCpLaf8aZOk:b-mdwhmg) ([raw file](https://github.com/batfish/batfish/blob/90ab55ff451f27f41970ff0eab85237d034c51ea/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/Route6FilterListMatchersImpl.java#L49)):*
> ```Java
>     @Override
>     public void describeTo(Description description) {
>       description.appendText(String.format("A RouteFilterList that rejects prefix: '%s'", _prefix));
> ```

`RouteFilterList` -> `Route6FilterList`

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 32 of 34 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java, line 20 at r2](https://reviewable.io/reviews/batfish/batfish/1426#-LD8b6az8EZvT7g7RLWY:-LD8fjbk20-xRDMUU0h2:b-1btnqc) ([raw file](https://github.com/batfish/batfish/blob/90ab55ff451f27f41970ff0eab85237d034c51ea/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixSpaceList.java#L20)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Your format string will provide a substitution of `true` or `false`. Was that your intention?
</blockquote></details>

That was by mistake, corrected... thanks for noticing

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/Route6FilterListMatchersImpl.java, line 49 at r2](https://reviewable.io/reviews/batfish/batfish/1426#-LD8c0BQDx15ssStxMUw:-LD8fvm-CN_dTTkzzGl_:b-1dcph8) ([raw file](https://github.com/batfish/batfish/blob/90ab55ff451f27f41970ff0eab85237d034c51ea/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/Route6FilterListMatchersImpl.java#L49)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

`RouteFilterList` -> `Route6FilterList`
</blockquote></details>

Corrected

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426)*
<!-- Sent from Reviewable.io -->

arifogel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426#-:-LD8gipv4hs3_PXWNno8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1426#-:-LD8oqK-EBh66N4TL9g3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
